File: plugins/data-ai-expert/skills/prompt-engineering-patterns/scripts/optimize-prompt.py
#!/usr/bin/env python3
from __future__ import annotations

"""
Prompt Optimization Script

一个轻量示例脚本，用于：
- 并行评估 prompt 变体
- 记录准确率、延迟、token 粗估
- 从若干启发式变体中选择更优版本

约束：
- 仅依赖 Python 标准库
- 默认使用内置 MockLLMClient 做演示
"""

import json
import math
import os
import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from statistics import fmean
from typing import Any


def percentile(values: list[float], pct: float) -> float:
    """计算线性插值百分位数。"""
    if not values:
        return 0.0

    ordered = sorted(values)
    if len(ordered) == 1:
        return float(ordered[0])

    rank = (pct / 100) * (len(ordered) - 1)
    lower = math.floor(rank)
    upper = math.ceil(rank)
    if lower == upper:
        return float(ordered[lower])

    weight = rank - lower
    return float(ordered[lower] + (ordered[upper] - ordered[lower]) * weight)


@dataclass(slots=True)
class TestCase:
    input: dict[str, Any]
    expected_output: str
    metadata: dict[str, Any] = field(default_factory=dict)


class PromptOptimizer:
    def __init__(self, llm_client, test_suite: list[TestCase], max_workers: int | None = None):
        self.client = llm_client
        self.test_suite = list(test_suite)
        self.results_history: list[dict[str, Any]] = []
        workers = max_workers or max(1, min(8, len(self.test_suite) or 1))
        self.executor = ThreadPoolExecutor(max_workers=workers)

    def shutdown(self) -> None:
        """关闭线程池。"""
        self.executor.shutdown(wait=True)

    def evaluate_prompt(
        self, prompt_template: str, test_cases: list[TestCase] | None = None
    ) -> dict[str, float]:
        """并行评估一个 prompt 模板。"""
        cases = test_cases or self.test_suite
        if not cases:
            return {
                "avg_accuracy": 0.0,
                "avg_latency": 0.0,
                "p95_latency": 0.0,
                "avg_tokens": 0.0,
                "success_rate": 0.0,
            }

        metrics = {
            "accuracy": [],
            "latency": [],
            "token_count": [],
            "success_rate": [],
        }

        def process_test_case(test_case: TestCase) -> dict[str, float]:
            started_at = time.perf_counter()

            try:
                prompt = prompt_template.format(**test_case.input)
                response = self.client.complete(prompt)
                response_text = "" if response is None else str(response)
                accuracy = self.calculate_accuracy(response_text, test_case.expected_output)
                success = 1.0 if response_text else 0.0
                token_count = float(len(prompt.split()) + len(response_text.split()))
            except Exception:
                accuracy = 0.0
                success = 0.0
                token_count = 0.0

            latency = time.perf_counter() - started_at
            return {
                "latency": float(latency),
                "token_count": token_count,
                "success_rate": success,
                "accuracy": accuracy,
            }

        results = list(self.executor.map(process_test_case, cases))
        for result in results:
            for key, value in result.items():
                metrics[key].append(value)

        return {
            "avg_accuracy": float(fmean(metrics["accuracy"])),
            "avg_latency": float(fmean(metrics["latency"])),
            "p95_latency": percentile(metrics["latency"], 95),
            "avg_tokens": float(fmean(metrics["token_count"])),
            "success_rate": float(fmean(metrics["success_rate"])),
        }

    def calculate_accuracy(self, response: str, expected: str) -> float:
        """用精确匹配 + 词重叠计算简易准确率。"""
        normalized_response = response.strip().lower()
        normalized_expected = expected.strip().lower()

        if normalized_response == normalized_expected:
            return 1.0

        response_words = set(normalized_response.split())
        expected_words = set(normalized_expected.split())
        if not expected_words:
            return 0.0

        overlap = len(response_words & expected_words)
        return overlap / len(expected_words)

    def optimize(self, base_prompt: str, max_iterations: int = 5) -> dict[str, Any]:
        """迭代优化 prompt。"""
        current_prompt = base_prompt
        best_prompt = base_prompt
        best_score = 0.0
        current_metrics: dict[str, float] | None = None

        for iteration in range(max_iterations):
            print(f"\nIteration {iteration + 1}/{max_iterations}")

            metrics = current_metrics or self.evaluate_prompt(current_prompt)
            print(
                f"Accuracy: {metrics['avg_accuracy']:.2f}, "
                f"Latency: {metrics['avg_latency']:.2f}s"
            )

            self.results_history.append(
                {
                    "iteration": iteration,
                    "prompt": current_prompt,
                    "metrics": metrics,
                }
            )

            if metrics["avg_accuracy"] > best_score:
                best_score = metrics["avg_accuracy"]
                best_prompt = current_prompt

            if metrics["avg_accuracy"] > 0.95:
                print("Achieved target accuracy!")
                break

            best_variation = current_prompt
            best_variation_metrics = metrics
            best_variation_score = metrics["avg_accuracy"]

            for variation in self.generate_variations(current_prompt, metrics):
                variation_metrics = self.evaluate_prompt(variation)
                if variation_metrics["avg_accuracy"] > best_variation_score:
                    best_variation = variation
                    best_variation_metrics = variation_metrics
                    best_variation_score = variation_metrics["avg_accuracy"]

            current_prompt = best_variation
            current_metrics = best_variation_metrics

            if best_variation_score > best_score:
                best_score = best_variation_score
                best_prompt = best_variation

        return {
            "best_prompt": best_prompt,
            "best_score": best_score,
            "history": self.results_history,
        }

    def generate_variations(
        self, prompt: str, current_metrics: dict[str, float]
    ) -> list[str]:
        """基于简单启发式生成候选变体。"""
        variations = [
            prompt + "\n\nProvide your answer in a clear, concise format.",
            "Let's solve this step by step.\n\n" + prompt,
            prompt + "\n\nVerify your answer before responding.",
        ]

        concise_prompt = self.make_concise(prompt)
        if concise_prompt != prompt:
            variations.append(concise_prompt)

        if "example" not in prompt.lower():
            variations.append(self.add_examples(prompt))

        # 准确率已经很高时，优先保留更短版本
        if current_metrics.get("avg_accuracy", 0.0) >= 0.8:
            variations = sorted(set(variations), key=len)
        else:
            variations = list(dict.fromkeys(variations))

        return variations[:3]

    def make_concise(self, prompt: str) -> str:
        """删除明显冗余的英语短语。"""
        replacements = [
            ("in order to", "to"),
            ("due to the fact that", "because"),
            ("at this point in time", "now"),
            ("in the event that", "if"),
        ]

        result = prompt
        for old, new in replacements:
            result = result.replace(old, new)
        return result

    def add_examples(self, prompt: str) -> str:
        """补一个最小示例。"""
        return f"""{prompt}

Example:
Input: Sample input
Output: Sample output
"""

    def compare_prompts(self, prompt_a: str, prompt_b: str) -> dict[str, Any]:
        """A/B 对比两个 prompt。"""
        print("Testing Prompt A...")
        metrics_a = self.evaluate_prompt(prompt_a)

        print("Testing Prompt B...")
        metrics_b = self.evaluate_prompt(prompt_b)

        return {
            "prompt_a_metrics": metrics_a,
            "prompt_b_metrics": metrics_b,
            "winner": "A" if metrics_a["avg_accuracy"] > metrics_b["avg_accuracy"] else "B",
            "improvement": abs(metrics_a["avg_accuracy"] - metrics_b["avg_accuracy"]),
        }

    def export_results(self, filename: str) -> None:
        """导出评估历史。"""
        output_dir = os.path.dirname(filename)
        if output_dir:
            os.makedirs(output_dir, exist_ok=True)
        with open(filename, "w", encoding="utf-8") as handle:
            json.dump(self.results_history, handle, indent=2, ensure_ascii=False)


def main() -> None:
    """本地演示入口。"""
    test_suite = [
        TestCase(input={"text": "This movie was amazing!"}, expected_output="Positive"),
        TestCase(input={"text": "Worst purchase ever."}, expected_output="Negative"),
        TestCase(
            input={"text": "It was okay, nothing special."},
            expected_output="Neutral",
        ),
    ]

    class MockLLMClient:
        def complete(self, prompt: str) -> str:
            if "amazing" in prompt.lower():
                return "Positive"
            if "worst" in prompt.lower():
                return "Negative"
            return "Neutral"

    optimizer = PromptOptimizer(MockLLMClient(), test_suite)
    try:
        base_prompt = "Classify the sentiment of: {text}\nSentiment:"
        results = optimizer.optimize(base_prompt)

        print("\n" + "=" * 50)
        print("Optimization Complete!")
        print(f"Best Accuracy: {results['best_score']:.2f}")
        print(f"Best Prompt:\n{results['best_prompt']}")

        optimizer.export_results("optimization_results.json")
    finally:
        optimizer.shutdown()


if __name__ == "__main__":
    main()

File: src/components/procedures/sources/prompt-engineering-patterns/optimize-prompt.ts
#!/usr/bin/env node
import { defineCliProcedure, procedureEntry } from "../../definition";
/**
 * Prompt Optimization Script
 *
 * Lightweight local demo for evaluating prompt variants, recording accuracy,
 * latency, token estimates, and selecting a better prompt from heuristic
 * variants. Uses only Node.js built-ins and the default MockLLMClient.
 */
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { performance } from "node:perf_hooks";

export const procedure = defineCliProcedure({
  id: "prompt-engineering-patterns-optimize-prompt",
  entry: procedureEntry(import.meta.url),
  description:
    "Prompt 优化演示脚本：使用内置示例对 prompt 变体进行精度、延迟、token 消耗的评估和迭代优化，输出优化结果 JSON。",
  owners: { skillIds: ["prompt-engineering-patterns"] },
  target: "scripts/optimize-prompt.mjs",
  runtime: "node",
  params: [
    {
      flag: "--output",
      type: "路径",
      description: "优化结果 JSON 输出路径（默认 optimization_results.json）",
      required: false,
    },
    {
      flag: "--overwrite",
      type: "",
      description: "允许覆盖已存在的优化结果输出；仅在确认目标文件可替换后使用",
      required: false,
    },
  ],

  exampleArgs: { args: [] },
});

function percentile(values: any, pct: any): any {
  if (values.length === 0) {
    return 0.0;
  }
  const ordered = [...values].sort((a: any, b: any) => a - b);
  if (ordered.length === 1) {
    return ordered[0];
  }
  const rank = (pct / 100) * (ordered.length - 1);
  const lower = Math.floor(rank);
  const upper = Math.ceil(rank);
  if (lower === upper) {
    return ordered[lower];
  }
  const weight = rank - lower;
  return ordered[lower] + (ordered[upper] - ordered[lower]) * weight;
}
function mean(values: any): any {
  return (
    values.reduce((total: any, value: any) => total + value, 0) / values.length
  );
}
class TestCase {
  expectedOutput: any;
  input: any;
  metadata: any;
  constructor(input: any, expectedOutput: any, metadata: any = {}) {
    this.input = input;
    this.expectedOutput = expectedOutput;
    this.metadata = metadata;
  }
}
class PromptOptimizer {
  client: any;
  maxWorkers: any;
  resultsHistory: any;
  testSuite: any;
  constructor(llmClient: any, testSuite: any, maxWorkers: any = null) {
    this.client = llmClient;
    this.testSuite = [...testSuite];
    this.resultsHistory = [];
    this.maxWorkers =
      maxWorkers ?? Math.max(1, Math.min(8, this.testSuite.length || 1));
  }
  shutdown(): any {}
  evaluatePrompt(promptTemplate: any, testCases: any = null): any {
    const cases = testCases ?? this.testSuite;
    if (cases.length === 0) {
      return {
        avg_accuracy: 0.0,
        avg_latency: 0.0,
        p95_latency: 0.0,
        avg_tokens: 0.0,
        success_rate: 0.0,
      };
    }
    const metrics: Record<string, any> = {
      accuracy: [],
      latency: [],
      token_count: [],
      success_rate: [],
    };
    const results = cases.map((testCase: any) =>
      this.processTestCase(promptTemplate, testCase),
    );
    for (const result of results) {
      for (const [key, value] of Object.entries(result)) {
        metrics[key].push(value);
      }
    }
    return {
      avg_accuracy: mean(metrics.accuracy),
      avg_latency: mean(metrics.latency),
      p95_latency: percentile(metrics.latency, 95),
      avg_tokens: mean(metrics.token_count),
      success_rate: mean(metrics.success_rate),
    };
  }
  processTestCase(promptTemplate: any, testCase: any): any {
    const startedAt = performance.now();
    let accuracy = 0.0;
    let success = 0.0;
    let tokenCount = 0.0;
    try {
      const prompt = formatTemplate(promptTemplate, testCase.input);
      const response = this.client.complete(prompt);
      const responseText =
        response === null || response === undefined ? "" : String(response);
      accuracy = this.calculateAccuracy(responseText, testCase.expectedOutput);
      success = responseText ? 1.0 : 0.0;
      tokenCount =
        prompt.split(/\s+/).filter(Boolean).length +
        responseText.split(/\s+/).filter(Boolean).length;
    } catch {
      accuracy = 0.0;
      success = 0.0;
      tokenCount = 0.0;
    }
    const latency = (performance.now() - startedAt) / 1000;
    return {
      latency,
      token_count: tokenCount,
      success_rate: success,
      accuracy,
    };
  }
  calculateAccuracy(response: any, expected: any): any {
    const normalizedResponse = response.trim().toLowerCase();
    const normalizedExpected = expected.trim().toLowerCase();
    if (normalizedResponse === normalizedExpected) {
      return 1.0;
    }
    const responseWords = new Set(
      normalizedResponse.split(/\s+/).filter(Boolean),
    );
    const expectedWords = new Set(
      normalizedExpected.split(/\s+/).filter(Boolean),
    );
    if (expectedWords.size === 0) {
      return 0.0;
    }
    let overlap = 0;
    for (const word of responseWords) {
      if (expectedWords.has(word)) {
        overlap += 1;
      }
    }
    return overlap / expectedWords.size;
  }
  optimize(basePrompt: any, maxIterations: any = 5): any {
    let currentPrompt = basePrompt;
    let bestPrompt = basePrompt;
    let bestScore = 0.0;
    let currentMetrics: any = null;
    for (let iteration = 0; iteration < maxIterations; iteration += 1) {
      console.log(`\nIteration ${iteration + 1}/${maxIterations}`);
      const metrics = currentMetrics ?? this.evaluatePrompt(currentPrompt);
      console.log(
        `Accuracy: ${metrics.avg_accuracy.toFixed(2)}, Latency: ${metrics.avg_latency.toFixed(2)}s`,
      );
      this.resultsHistory.push({
        iteration,
        prompt: currentPrompt,
        metrics,
      });
      if (metrics.avg_accuracy > bestScore) {
        bestScore = metrics.avg_accuracy;
        bestPrompt = currentPrompt;
      }
      if (metrics.avg_accuracy > 0.95) {
        console.log("Achieved target accuracy!");
        break;
      }
      let bestVariation = currentPrompt;
      let bestVariationMetrics = metrics;
      let bestVariationScore = metrics.avg_accuracy;
      for (const variation of this.generateVariations(currentPrompt, metrics)) {
        const variationMetrics = this.evaluatePrompt(variation);
        if (variationMetrics.avg_accuracy > bestVariationScore) {
          bestVariation = variation;
          bestVariationMetrics = variationMetrics;
          bestVariationScore = variationMetrics.avg_accuracy;
        }
      }
      currentPrompt = bestVariation;
      currentMetrics = bestVariationMetrics;
      if (bestVariationScore > bestScore) {
        bestScore = bestVariationScore;
        bestPrompt = bestVariation;
      }
    }
    return {
      best_prompt: bestPrompt,
      best_score: bestScore,
      history: this.resultsHistory,
    };
  }
  generateVariations(prompt: any, currentMetrics: any): any {
    let variations: any[] = [
      `${prompt}\n\nProvide your answer in a clear, concise format.`,
      `Let's solve this step by step.\n\n${prompt}`,
      `${prompt}\n\nVerify your answer before responding.`,
    ];
    const concisePrompt = this.makeConcise(prompt);
    if (concisePrompt !== prompt) {
      variations.push(concisePrompt);
    }
    if (!prompt.toLowerCase().includes("example")) {
      variations.push(this.addExamples(prompt));
    }
    if ((currentMetrics.avg_accuracy ?? 0.0) >= 0.8) {
      variations = [...new Set(variations)].sort(
        (a: any, b: any) => a.length - b.length,
      );
    } else {
      variations = [
        ...new Map(
          variations.map((variation: any) => [variation, variation]),
        ).values(),
      ];
    }
    return variations.slice(0, 3);
  }
  makeConcise(prompt: any): any {
    const replacements: any[] = [
      ["in order to", "to"],
      ["due to the fact that", "because"],
      ["at this point in time", "now"],
      ["in the event that", "if"],
    ];
    let result = prompt;
    for (const [oldValue, newValue] of replacements) {
      result = result.split(oldValue).join(newValue);
    }
    return result;
  }
  addExamples(prompt: any): any {
    return `${prompt}

Example:
Input: Sample input
Output: Sample output
`;
  }
  comparePrompts(promptA: any, promptB: any): any {
    console.log("Testing Prompt A...");
    const metricsA = this.evaluatePrompt(promptA);
    console.log("Testing Prompt B...");
    const metricsB = this.evaluatePrompt(promptB);
    return {
      prompt_a_metrics: metricsA,
      prompt_b_metrics: metricsB,
      winner: metricsA.avg_accuracy > metricsB.avg_accuracy ? "A" : "B",
      improvement: Math.abs(metricsA.avg_accuracy - metricsB.avg_accuracy),
    };
  }
  exportResults(filename: any, overwrite: any = false): any {
    assertOutputWritable(filename, overwrite);
    const outputDir = dirname(filename);
    if (outputDir && outputDir !== ".") {
      mkdirSync(outputDir, { recursive: true });
    }
    writeFileSync(
      filename,
      JSON.stringify(this.resultsHistory, null, 2),
      "utf8",
    );
  }
}
export function assertOutputWritable(
  outputPath: any,
  overwrite: any = false,
): any {
  if (existsSync(outputPath) && !overwrite) {
    throw new Error(
      `output file already exists: ${outputPath}; pass --overwrite only after confirming it can be replaced`,
    );
  }
}
function formatTemplate(template: any, values: any): any {
  return template.replace(/\{([^{}]+)\}/g, (match: any, key: any) => {
    if (!(key in values)) {
      throw new Error(`Missing template value: ${key}`);
    }
    return String(values[key]);
  });
}
function usage(): any {
  return `Prompt optimization demo.

Usage: node scripts/optimize-prompt.mjs [options]

Options:
  --output <path>   Write optimization history JSON to this path (default: optimization_results.json)
  --overwrite       Replace an existing output file after confirmation
  --help            Show this help
`;
}
export function parseArgs(argv: readonly string[]): any {
  const args: Record<string, any> = {
    output: "optimization_results.json",
    overwrite: false,
    help: false,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--help" || arg === "-h") {
      args.help = true;
      continue;
    }
    if (arg === "--overwrite") {
      args.overwrite = true;
      continue;
    }
    if (arg === "--output") {
      const value = argv[index + 1];
      if (value == null || value.startsWith("--")) {
        throw new Error("--output requires a value");
      }
      args.output = value;
      index += 1;
      continue;
    }
    throw new Error(`unrecognized argument: ${arg}`);
  }
  return args;
}
export function main(argv: readonly string[]): any {
  const args = parseArgs(argv);
  if (args.help) {
    console.log(usage());
    return 0;
  }
  const testSuite: any[] = [
    new TestCase({ text: "This movie was amazing!" }, "Positive"),
    new TestCase({ text: "Worst purchase ever." }, "Negative"),
    new TestCase({ text: "It was okay, nothing special." }, "Neutral"),
  ];
  const mockLlmClient: Record<string, any> = {
    complete(prompt: any) {
      const normalized = prompt.toLowerCase();
      if (normalized.includes("amazing")) {
        return "Positive";
      }
      if (normalized.includes("worst")) {
        return "Negative";
      }
      return "Neutral";
    },
  };
  const optimizer = new PromptOptimizer(mockLlmClient, testSuite);
  try {
    const basePrompt = "Classify the sentiment of: {text}\nSentiment:";
    const results = optimizer.optimize(basePrompt);
    console.log(`\n${"=".repeat(50)}`);
    console.log("Optimization Complete!");
    console.log(`Best Accuracy: ${results.best_score.toFixed(2)}`);
    console.log(`Best Prompt:\n${results.best_prompt}`);
    optimizer.exportResults(args.output, args.overwrite);
  } finally {
    optimizer.shutdown();
  }
}
